[WGSL] Floating point -> integer conversions should be bounds checked

tadeuzagallo opened the pull request:
#### a77d8458cec597d4cea6086410967c2bc118ee3d
<pre>
[WGSL] Floating point -&gt; integer conversions should be bounds checked
<a href="https://bugs.webkit.org/show_bug.cgi?id=283980">https://bugs.webkit.org/show_bug.cgi?id=283980</a>
<a href="https://rdar.apple.com/140858383">rdar://140858383</a>

Reviewed by Mike Wyrzykowski.

Introduce helper function to avoid undefined behavior in float to integer
conversions.

* LayoutTests/fast/webgpu/ftoi-expected.txt: Added.
* LayoutTests/fast/webgpu/ftoi.html: Added.
* Source/WebGPU/WGSL/Metal/MetalFunctionWriter.cpp:
(WGSL::Metal::FunctionDefinitionWriter::emitNecessaryHelpers):
(WGSL::Metal::FunctionDefinitionWriter::visit):
* Source/WebGPU/WGSL/TypeCheck.cpp:
(WGSL::TypeChecker::chooseOverload):
* Source/WebGPU/WGSL/WGSLShaderModule.h:
(WGSL::ShaderModule::usesFtoi const):
(WGSL::ShaderModule::setUsesFtoi):

Canonical link: <a href="https://commits.webkit.org/288599@main">https://commits.webkit.org/288599@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/91c0bdd884a5bbacb3b2dd30819d8b9b34d20e89

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/83894 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/3512 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/38195 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/88968 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/34902 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/3603 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/11479 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/65253 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/23087 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/86940 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/2671 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/76226 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/45545 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/2600 "Passed tests") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/64/builds/30447 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/33951 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/73590 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/63/builds/31202 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/90344 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/11159 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/8068 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/73700 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/11383 "Built successfully") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/8/builds/72055 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/72918 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/18028 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/17200 "Passed tests") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/119/builds/15890 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/2493 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/11111 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/16583 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/10959 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/14435 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/12731 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

mwyrzykowski commented on Source/WebGPU/WGSL/Metal/MetalFunctionWriter.cpp:
```diff
@@ -527,6 +527,19 @@ void FunctionDefinitionWriter::emitNecessaryHelpers()
+        {
+            IndentationScope scope(m_indent);
+            m_stringBuilder.append(m_indent, "auto min = metal::numeric_limits<T>::min();\n"_s,
+                m_indent, "auto max = metal::numeric_limits<T>::max();\n"_s,
```
Oh I just realized these are `metal::numeric_limits` variants, please ignore the earlier comment

tadeuzagallo replied:
It's also operating on the target type, which is the integer type, so I think it's ok. But it won't handle booleans right now, so I have to fix that

mwyrzykowski replied:
Yes this is great, sorry ignore my earlier comment

tadeuzagallo replied:
I implemented it for booleans too, but then I checked and it doesn't seem like it's UB to convert any float values to bool. To confirm I tested many different values with UBSan and no issues were raised, so I'll leave it out for now.

mwyrzykowski reviewed (changes requested):
I think the test regressions are legitimate - https://ews-build.s3-us-west-2.amazonaws.com/macOS-Sequoia-Debug-WK2-Tests-EWS/f83960ff-7008/results.html

```
program_source:21:23: error: use of undeclared identifier '__wgslFtoi'
    global0[__wgslMin(__wgslFtoi(64), (268435455u - 1u))] = 287454020u;
```

mwyrzykowski commented on Source/WebGPU/WGSL/Metal/MetalFunctionWriter.cpp:
```diff
@@ -527,6 +527,19 @@ void FunctionDefinitionWriter::emitNecessaryHelpers()
+            m_indent, "static T __wgslFtoi(S value)\n"_s,
+            m_indent, "{\n"_s);
+        {
+            IndentationScope scope(m_indent);
```
Proposed change:
```
            m_stringBuilder.append(
                m_indent, "if constexpr (is_same_v<T, S>) {\n"_s,
                m_indent, m_indent, "return value;\n"_s,
                m_indent, "} else if constexpr (is_floating_point_v<S>) {\n"_s,
                m_indent, m_indent, "auto min = metal::numeric_limits<T>::min();\n"_s,
                m_indent, m_indent, "auto max = metal::numeric_limits<T>::max();\n"_s,
                m_indent, m_indent, "return T(metal::clamp(value, S(min), S(max)));\n"_s,
                m_indent, "} else if constexpr (is_unsigned_v<T> && is_signed_v<S> && sizeof(S) <= sizeof(T)) {\n"_s,
                m_indent, m_indent, "return T(metal::max(S(0), value));\n"_s,
                m_indent, "} else if constexpr (is_unsigned_v<T> && is_signed_v<S>) {\n"_s,
                m_indent, m_indent, "auto max = metal::numeric_limits<T>::max();\n"_s,
                m_indent, m_indent, "return T(metal::clamp(value, S(0), S(max)));\n"_s,
                m_indent, "} else if constexpr (is_signed_v<T> && is_signed_v<S>) {\n"_s,
                m_indent, m_indent, "auto min = metal::numeric_limits<T>::min();\n"_s,
                m_indent, m_indent, "auto max = metal::numeric_limits<T>::max();\n"_s,
                m_indent, m_indent, "return T(metal::clamp(value, S(min), S(max)));\n"_s,
                m_indent, "} else {\n"_s,
                m_indent, m_indent, "return T(value);\n"_s,
                m_indent, "}\n"_s);
```

because this is getting called with S = int and T = unsigned and I think `unsigned(clamp(value, int(0), int(0xFFFFFFFF))` is undefined behavior, at least I see different behavior on https://playground.babylonjs.com/#18WTCK after this change (textures appear white instead of brown)

tadeuzagallo replied:
Thanks for catching it. The issue was that I didn't mean to emit `__wgslFtoi` when T was neither `f32` or `f16`, but `satisfies(AbstractInt, Constraints::Float)` is true (since we can promote then `AbstractInt` to `Float`), so that resulted in unexpected calls to ftoi. I ended up changing it to `satisfies(T, Constraints::Float) && !satisfies(T, Constraints::AbstractInt)`, which isn't ideal, but for now we don't have a convenient way of checking if a type satisfies a constraint without promoting it (the alternative would be checking `T == F16 || T == F32`, but that has other drawbacks, like not handling references)

tadeuzagallo committed:
[WGSL] Floating point -> integer conversions should be bounds checked
https://bugs.webkit.org/show_bug.cgi?id=283980
rdar://140858383

Reviewed by Mike Wyrzykowski.

Introduce helper function to avoid undefined behavior in float to integer
conversions.

* LayoutTests/fast/webgpu/ftoi-expected.txt: Added.
* LayoutTests/fast/webgpu/ftoi.html: Added.
* Source/WebGPU/WGSL/Metal/MetalFunctionWriter.cpp:
(WGSL::Metal::FunctionDefinitionWriter::emitNecessaryHelpers):
(WGSL::Metal::FunctionDefinitionWriter::visit):
* Source/WebGPU/WGSL/TypeCheck.cpp:
(WGSL::TypeChecker::chooseOverload):
* Source/WebGPU/WGSL/WGSLShaderModule.h:
(WGSL::ShaderModule::usesFtoi const):
(WGSL::ShaderModule::setUsesFtoi):

Canonical link: https://commits.webkit.org/288599@main
